### ISSUE TITLE
Error using PassiveAgressiveClassifier on Text Data

### ISSUE THREAD

**Issue by Yasmen-Wahba** [online-ml/river#1106]
Hello, I'm trying to use the PassiveAgressive classifier on my confidential text dataset. I'm getting the following error on Google Colab Environment=>
**_TypeError: unsupported operand type(s) for -: 'str' and 'int'_**

This is the code I'm using, **X** is the text dataframe and **y** is a Series describing the labels (i.e., category)

**Python 3.9.15**:
**Windows**:

## Steps/code to reproduce
```
model = linear_model.PAClassifier(C=0.1, mode=1)
pipe_pac = Pipeline(('vectorizer',BagOfWords(lowercase=True,ngram_range=(1, 3))),('pac',model))
X= df['Text']
y= df['CATEGORY']

X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.3, random_state = 0)
newdf = pd.concat([X_train, pd.DataFrame(y_train)], axis=1)

data = list(zip(*map(newdf.get, newdf)))

for text,label in data:
    pipe_pac = pipe_pac.learn_one(text,label)

```

**Comment by MaxHalford** [comment]
Hey @Yasmen-Wahba. You can't provide an example dataset, so I asked ChatGPT to generate me one:

![Screenshot 2022-12-06 at 09 16 09](https://user-images.githubusercontent.com/8095957/205857066-525da806-df7c-46f3-a632-e3f6e8a72178.png)

This is the code I got:

```py
import pandas as pd
import random

data = {
    'text': [],
    'is_positive': []
}

# generate 100 random sentences and booleans
for i in range(100):
    sentence = "This is a "
    if random.random() < 0.5:
        sentence += "positive"
        is_positive = True
    else:
        sentence += "negative"
        is_positive = False
    sentence += " sentence."
    
    data['text'].append(sentence)
    data['is_positive'].append(is_positive)

df = pd.DataFrame(data)
```

Here are the first 5 rows:

|    | text                         | is_positive   |
|---:|:-----------------------------|:--------------|
|  0 | This is a negative sentence. | False         |
|  1 | This is a negative sentence. | False         |
|  2 | This is a positive sentence. | True          |
|  3 | This is a negative sentence. | False         |
|  4 | This is a positive sentence. | True          |

Does this look like what you have? Note that I used booleans as labels because `PAClassifier` is a binary classifier, not a multi-class classifier.

Then I ran this code (I had to slightly edit yours):

```py
from river import compose
from river import feature_extraction
from river import linear_model
from sklearn import model_selection

model = linear_model.PAClassifier(C=0.1, mode=1)
pipe_pac = compose.Pipeline(
    ('vectorizer', feature_extraction.BagOfWords(lowercase=True, ngram_range=(1, 3))),
    ('pac', model)
)

X = df['text']
y = df['is_positive']

X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, test_size=0.3, random_state=0)
newdf = pd.concat([X_train, pd.DataFrame(y_train)], axis=1)

data = list(zip(*map(newdf.get, newdf)))

for text,label in data:
    pipe_pac = pipe_pac.learn_one(text,label)
```

Everything is working fine on my end.

**Comment by MaxHalford** [comment]
If you have a link to the implementation you were using, or a research paper, then we could implement it for you. Or if you feel like contributing, that would be even better :)

**Comment by smastelini** [comment]
How about using something like [this](https://riverml.xyz/0.14.0/api/multiclass/OneVsOneClassifier/)?

**Comment by Yasmen-Wahba** [comment]
> If you have a link to the implementation you were using, or a research paper, then we could implement it for you. Or if you feel like contributing, that would be even better :)
> 
> Note that if you were using scikit-learn's implementation, I believe they're using a one-vs-all strategy to make it work for multi-class problems. You can do that with River thanks to the `multiclass` module.

Yes, I was using Scikit-learn's implementation!

**Comment by Yasmen-Wahba** [comment]
> How about using something like [this](https://riverml.xyz/0.14.0/api/multiclass/OneVsOneClassifier/)?

I was just about to try that as well as OutputcodeClassifier

**Comment by Yasmen-Wahba** [comment]
I just noticed a typo here: https://riverml.xyz/0.14.0/api/multiclass/OutputCodeClassifier/
`ooc = OutputCodeClassifier( `==> should be `ooc = multiclass.OutputCodeClassifier(`
I know it's trivial but just to avoid someone opening an issue for it :)

**Comment by MaxHalford** [comment]
Ah good catch. Do you mind fixing that with a pull request? That way you get the to be the one who fixes it in the git history :)

**Comment by Yasmen-Wahba** [comment]
I was just wondering if there is a way to use river to classify a 2-level hierarchy, that is passing the labels as a 2-d array (e.g., product, sub-product).

**Comment by MaxHalford** [comment]
> I was just wondering if there is a way to use river to classify a 2-level hierarchy, that is passing the labels as a 2-d array (e.g., product, sub-product).

Sadly not yet no :)

It is on the roadmap though to frame multi-class classification as a tree of binary classifiers. But it will take a lot of work to get that done.
